### ISSUE TITLE
What is the future for the alpha simulator?

### ISSUE THREAD

**Issue by afberendsen** [open-simh/simh#222]
I used to work for DEC/Compq/HP supporting Alpha based solutions.

There are few Alpha simulators on the market, including SimH.

What is the future for the SimH Alpha simulator? 

I am more than keen to help improve it. 

At high level, the Alpha had a similar behaviour as from the VAX-8600.

**Comment by markpizz** [comment]
The simh Alpha simulator only simulates the CPU and not a complete system.  I'm not sure that anyone has access to sufficient detailed information to create all the missing pieces of a working system.  If you do, please jump in and help make some progress.

**Comment by afberendsen** [comment]
Yeah...It will take me some time to learn and build enough knowledge on how to do it. I was hoping to help someone with more knowledge until I can be more productive.

**Comment by markpizz** [comment]
Over the years that the alpha CPU has been in the codebase, a couple of people have expressed interest in doing something, but we haven't seen or heard of any progress.  Significant detailed information and likely access to source code would probably be needed to get to a working solution.

**Comment by pkoning2** [comment]
There are other simulators that are much more complete, such as "axpbox".  Unfortunately that one is GPL so we can't directly draw from it, never mind the fact that it's written in C++.

**Comment by afberendsen** [comment]
What sort of documents/information will be relevant to help?

**Comment by afberendsen** [comment]
...and another ones with the titles:
- _AlphaServers GS80/160/1280 System Technologies_
- _AlphaServer DS1x/DS2x/ES45 System Technologies_

**Comment by afberendsen** [comment]
um...It can sound like a stupid question, but using the *BSD source codes will not provide the required information?

FreeBSD 6.0 supported all Alpha platforms up to ES40: https://www.freebsd.org/releases/6.0R/hardware-alpha/
NetBSD support all the first models: http://wiki.netbsd.org/ports/alpha/
OpenBSD supports a large number of systems: https://www.openbsd.org/alpha.html

**Comment by pkoning2** [comment]
Reverse engineering hardware behavior from *BSD sources is an option, but not nearly as good as using specs.  So if specs can be found that would be the best answer.
Also, being bug-for-bug compatible with some BSD isn't necessarily the best option.  For example, axpbox is documented as supporting VMS, and also NetBSD with the exception of networking.  Does that mean the Ethernet simulation has a bug?  Or does it mean NetBSD does?  Hard to know, without the specs.

**Comment by afberendsen** [comment]
That's true. But, at the same time, it is a start. 

Since there are three different BDS options to consider, and if each BSD has its own NIC driver, it is more likely that it will work. 

After having something running, and if problems are found, contacting HP for help would a natural path.

Better to start with what is available and improve as the development goes.

**Comment by markpizz** [comment]
I'd be surprised if all of the BSD versions didn't have the same axp drivers....

**Comment by markpizz** [comment]
All reasonable ideas.

The starting point (before any OS level functionality can be done) is to make it through the boot ROM images which are certainly available online somewhere.  This, by itself, will be a challenging hill to climb.  You and others are welcome to try and jump into these waters whenever you are inspired.

**Comment by afberendsen** [comment]
Well, after checking all the available Alpha emulators, all of them came from the same root program. The configuration files are remarkably similar, and the general coding structure is similar (strings). I will try to mix the current SimH Alpha, plus some ideas from other sources to have a basic running environment for the SRM. At this stage I do not need to consider any I/O devices besides the serial port. 

Also, I found tons of good DEC/Compaq documentation covering very low-level system architecture, chipset internal workings. 

I Should give it a try. I have loads of free time to use :)

**Comment by pkoning2** [comment]
That's great news!

I just thought of another drawback of using device drivers as a substitute for device specs: those only tell you about the subset of device features that the driver happens to use.  To pick an older example: looking at most PDP-11 OS device drivers will tell you nothing about the behavior of partial block writes, since most OS don't ever do that.  But RT-11 does and has a specific requirement for how it behaves.  A device emulation writer has to be aware of that to get it to work.  If you didn't know that obscure bit of information, and didn't have RT-11 driver sources to examine, you might have a hole in the emulation.

**Comment by afberendsen** [comment]
Very good points.
I will do my best with what I have with me.

If I am stuck somewhere, I suppose I can use the source code from the ES40 and axpbox projects.  Is it legal?

I do not want to start with an ES40, since it is a multi CPU architecture and there are many other challenges to deal with.

From the list of Alpa systems, the Alpha Sever 400 seems to be the most simple platform.

I managed to unpack about 100GB of documents from my very old backups. Also, The Internet Archive proved to be an amazing source for old documents.

**Comment by pkoning2** [comment]
On APXbox etc. source code: please be careful.  That code is GPL, which means you cannot use it, or parts of it, in SIMH.

Using code as a form of documentation, and then writing your own code that is not a copy of the GPL code, is different.  But the problem is to do that in a way that doesn't expose you to claims of having copied the GPL code.

Using documents rather than GPL code is a better and safer way to go.  

One other option would be to work with two people: one to turn what you can learn from GPL code into specifications, the other to write new non-GPL code using those specifications as input.  But to do it all as a single person is hard to get right.

**Comment by afberendsen** [comment]
I see..."a la" creating a new BIOS from the IBM PC BIOS :)

**Comment by pkoning2** [comment]
Yes, whenever you make use of someone else's work, you need to consider just how you are using it and whether you have permission to do that.  
For open source, it's easy to get confused and think "open" means "no restrictions".  But that is not true (with the very rare exception of actual "released into the public domain" code).  Instead, such code is covered by a license that restricts what you can do with it.  The GPL license is particularly tricky because it "infects" other code -- if you merge GPL with something else, by its rules the GPL then applies to the entire combined work.  And as a rule, other non-GPL open source projects don't want that, so they refuse to accept any contributions that would bring in GPL bits.
